File: algorithms/medium/product_of_array_except_self.ex
defmodule Solution do
  @spec product_except_self(nums :: [integer]) :: [integer]
  def product_except_self(nums) do
    prefix = nums |> product_prefix(1)
    suffix = nums |> Enum.reverse() |> product_prefix(1) |> Enum.reverse()
    Enum.zip(prefix, suffix) |> Enum.map(fn {a, b} -> a * b end)
  end

  def product_prefix([], _prefix), do: []

  def product_prefix([x | xs], prefix) do
    [prefix | product_prefix(xs, prefix * x)]
  end
end

File: algorithms/medium/product_of_array_except_self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut ans = Vec::with_capacity(nums.len());
        let mut prefix = 1;
        for &n in &nums {
            ans.push(prefix);
            prefix *= n;
        }

        let mut suffix = 1;
        for (i, n) in nums.iter().enumerate().rev() {
            ans[i] *= suffix;
            suffix *= nums[i];
        }
        ans
    }
}
